### PR TITLE
Wrapping compiler builtins

### PR DESCRIPTION
**Pull request by stoklund**
As [recently proposed on Zephyr-devel](https://lists.zephyrproject.org/g/devel/topic/31422932), this PR implements wrapper functions for a number of commonly used compiler builtins.

- CMake is used to detect which builtins are supported by the target compiler. A `<ccfeatures.h>` header file is generated.
- The new functions are provided in `<misc/math_extras.h>` with inline implementations that are either portable C or using compiler builtins when available.

I also replaced the use of naked overflow arithmetic builtins throughout the code. I have not yet done the same for CLZ/CTZ.

There are more candidates for wrapping like this:

- The `find_lsb_set()` and `find_msb_set()` functions could be implemented in terms of `__builtin_ffs()` and the existing bit-counting functions. It may also make sense to provide these in `size_t` variants, depending on how they are used.
- Byte swapping operations as currently provided by `<misc/byteorder.h>`.

### CODE REVIEW

**Review by andyross — CHANGES_REQUESTED** [review]
Not really loving the way this inserts cmake into the lowest level of the build.  Build systems come and go like leaves on the wind.   And testing these all with per-app configure-style build-a-test-c-program seems terribly wasteful for a project that has IIRC 1700 such apps built for every CI pass.  And... what does this give us that we don't already have?  We know very well already which compilers support these builtins statically, there's zero need to probe this at build time.

FWIW: the abstracted API and generic fallback implementations look great to me.  It's the selection via complicated build magic that seems fragile and needless.

**Review by tyrantlizard77 — COMMENTED** [review]
Overall LGTM, at least the interface definition and C implementations.
Is there a way to have two cases in math_extras testcase.yaml, one of which forces the C implementations to be exercised?

**Comment by stoklund** [comment]
> We know very well already which compilers support these builtins statically, there's zero need to probe this at build time.

I'm probably missing something obvious, but could you explain how that would work in practice?

I don't think compilers are publishing tables of which release first supported a given builtin, so you have to go through old release notes to figure that out. That is work, but doable of course.

When you test for compiler versions using macros like `__GNUC__`, so have to look out for compilers like Clang and ICC that masquerade as a certain GCC version by also defining this macro. There are cases of [ICC claiming to be compatible with a certain GCC version, but with some builtins missing](https://software.intel.com/en-us/forums/intel-c-compiler/topic/758013).

Clang has a `__has_builtin()` macro which does exactly what we need, but unfortunately only Clang has that, AFAIK.

I am also a little concerned about various toolchains from vendors. These are often based on GCC or Clang, but with a number of patches applied. This means that builtins can come and go while the compiler still claims to be some major version of GCC/Clang.

> Is there a way to have two cases in math_extras testcase.yaml, one of which forces the C implementations to be exercised?

I added a commit which does this.

File: include/misc/math_extras_impl.h [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+	} else {
+		return u32_count_leading_zeros(x >> 32);
+	}
+}
+
+static inline int u32_count_trailing_zeros(u32_t x)
+{
+	int b = 0;
+
+	while (b < 32 && (x & 1) == 0)
```
Comment (jukkar) [inline include/misc/math_extras_impl.h]: Missing {} here

File: include/misc/math_extras_impl.h [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+		b++, x >>= 1;
+	return b;
```
Comment (jukkar) [inline include/misc/math_extras_impl.h]: empty line needed before the return statement

**Comment by pfalcon** [comment]
> The new functions are provided in <misc/math_extras.h>

I'd say, I find such a name unexpected. math.h is a well-known header which provides functions like sin() and cos(). Now I'd find strange to find functions like CLZ in math_extras.h. That said, I don't really have suggestion for better name (I'd apparently emphasize that it's wrappers for compiler builtin functions, with a name like cc_wrappers.h).

>  And testing these all with per-app configure-style build-a-test-c-program seems terribly wasteful

+1 for the concern of calling gazillion of check_c_source_compiles() which apparently spawn a subprocess each time.

File: Kconfig.zephyr [hunk 3]
```diff
@@ -250,6 +250,14 @@ config COMPILER_OPT
 	  and can be used to change compiler optimization, warning and error
 	  messages, and so on.
 
+config AVOID_COMPILER_BUILTINS
+	bool "Avoid using compiler builtin functions"
+	help
+	  When this option is set, Zephyr will avoid using the compiler's
+	  builtin functions for extra math operations. Portable C
```
Comment (SebastianBoe) [inline Kconfig.zephyr]: > for extra math operations.

The option name implies something else. It implies we would avoid built-in memcmp etc. as well.

Perhaps the option name should be extended to disambiguate?

File: cmake/ccfeatures.cmake [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  }
+" HAVE___BUILTIN_CLZLL)
+
+check_c_source_compiles("
+  int main(void) {
+    return __builtin_ctz(2);
+  }
+" HAVE___BUILTIN_CTZ)
+
+check_c_source_compiles("
```
Comment (SebastianBoe) [inline cmake/ccfeatures.cmake]: This must be integrated with the ZEPHYR_TOOLCHAIN_CAPABILITY_CACHE to reach an acceptable cmake configure performance.

**Review by SebastianBoe — DISMISSED** [review]

**Comment by SebastianBoe** [comment]
> Not really loving the way this inserts cmake into the lowest level of the build.

I might misunderstand you, but be aware that we are already checking for compiler flag compatibility at build time. This is behaviour inherited from the Linux kernel (cc-option), and presumably exists to easily build the kernel with a wide range of gcc versions.

The use of #cmakedefine and configure_file is also in use, e.g. with version.h, and has proven to work well. It should be noted, that migrating to a different build system would only need porting of the *.h.in file, not all users of the .h file.

File: include/misc/math_extras_impl.h (hunk 2 above)
Comment (stoklund) [reply include/misc/math_extras_impl.h]: Thanks, I'll fix these lints.

**Comment by olofj** [comment]
@andyross

Linux does this at build time by detecting the compiler being used instead of having the build system do it. I.e. there's a <linux/compiler.h> that detects compiler based on defines. Nothing quite as elaborate is needed here yet, but the basic approach seems useful:

https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/linux/compiler_types.h#n58

I.e. #ifdef + per-compiler sub-include, and then in that include key off of version if needed.

Would that be an approach you'd support?

**Comment by stoklund** [comment]
> > The new functions are provided in <misc/math_extras.h>
> 
> I'd say, I find such a name unexpected. math.h is a well-known header which provides functions like sin() and cos(). Now I'd find strange to find functions like CLZ in math_extras.h. That said, I don't really have suggestion for better name (I'd apparently emphasize that it's wrappers for compiler builtin functions, with a name like cc_wrappers.h).

Heh. I happen to be a mathematician, and I find the C standard committee's idea of "math" rather limiting ;-)

But seriously, the name comes from the LLVM and Mozilla projects which have `MathExtras.h` and `MathAlgorithms.h` headers respectively. The idea behind these names is that the headers provide little functions that just operate on numbers. The functions don't have any context other than numbers, hence "math".

An alternative is `<misc/int_extras.h>` since these are all missing integer operations. (Rust provides all of these ops [as part of their integer types](https://doc.rust-lang.org/std/primitive.u32.html)).

IMHO, the fact that many of these functions are implemented in terms of compiler builtins should just be an implementation detail. Somebody using this header is not going to care if they are calling portable C functions or builtins.

File: Kconfig.zephyr (hunk 3 above)
Comment (stoklund) [reply Kconfig.zephyr]: Yeah, I agree.

I'm not even sure if this should be a Kconfig option or just a feature macro (with a long name) that can be defined before including the header. It's really only useful for the unit test which wants to test the portable implementations.

**Comment by andyross** [comment]
FWIW, the cmake criticism was mostly aesthetics.  I'm not really interested in litgating that per se, the point is that I'd prefer the lowest level of our headers be free of tool interactions.  C code should be C code.

The focused complaint was the runtime testing, which is just a bad idea given the breadth and structure of our test suite.

The broader complaint is that we already have a framework for specifying toolchain configuration statically via ZEPHYR_TOOLCHAIN_VARIANT and the toolchain.h header tree, and we should be leveraging that if at all possible.  I just don't see the need to try to detect this at runtime, do we even **HAVE** toolchains that support these functions piecewise?  Seems like all we really care about is "gcc" and "other", and can just handle things the way we already do, no?

**Comment by stoklund** [comment]
Thanks for the reviews. I've pushed an updated version which:

* Doesn't use CMake to detect the presence of compiler builtins at configure time. Instead, it uses a similar method to the Linux kernel to detect builtins at compile time. The existing `<toolchain.h>` header now provides a `HAS_BUILTIN(x)` macro for the purpose.
* Doesn't add anything to `Kconfig.zephyr`. The new header file just detects the presence of a `PORTABLE_MISC_MATH_EXTRAS` feature macro to disable the use of compiler builtins for testing purposes.
* Fixes nits found by reviewers and tools.

File: include/misc/math_extras_impl.h [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+	return __builtin_mul_overflow(a, b, result);
+}
+#else /* !use_builtin(__builtin_mul_overflow) */
+static inline bool u32_mul_overflow(u32_t a, u32_t b, u32_t *result)
+{
+	u32_t c = a * b;
+
+	*result = c;
+
+	return a > 0 && c / a != b;
```
Comment (mped-oticon) [inline include/misc/math_extras_impl.h]: nit: Parens would be nice to make precedence clear.

**Review by mped-oticon — APPROVED** [review]
LGTM. This a welcome change that adds toolchain flexibility and rigour.

File: include/misc/math_extras_impl.h (hunk 5 above)
Comment (andyross) [reply include/misc/math_extras_impl.h]: Alternative nit: "n != 0" better captures the logic.  You're avoiding the divide by zero edge case here.  My first thought on seeing that > was that you must have a overflow/signedness bug since a is unsigned.

File: include/misc/math_extras_impl.h [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+{
+	return x == 0 ? 32 : __builtin_ctz(x);
+}
+#else /* !use_builtin(__builtin_ctz) */
+static inline int u32_count_trailing_zeros(u32_t x)
+{
+	int b = 0;
+
+	while (b < 32 && (x & 1) == 0) {
+		b++, x >>= 1;
```
Comment (andyross) [inline include/misc/math_extras_impl.h]: So... frankly I don't personally care, but that use of the comma operator is needless and I'm all but certain someone's coding style document (@ceolin can tell you all about the world he's been stuck in) disallows this.   Actually given the structure, maybe this should just be a `for()`  (where, heh, you could use the comma in a well-justified way because it has to fit in the final expression).

File: include/misc/math_extras_impl.h [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+	}
+
+	return b;
+}
+#endif /* use_builtin(__builtin_ctz) */
+
+#if use_builtin(__builtin_ctzll)
+static inline int u64_count_trailing_zeros(u64_t x)
+{
+	return x == 0 ? 64 : __builtin_ctzll(x);
```
Comment (andyross) [inline include/misc/math_extras_impl.h]: Genuine confusion here.  Why are we making the semantics (for an argument of zero) of our own "count_trailing_zeros" function...  **DIFFERENT** than the "ctz" builtin that we're supposed to be wrapping?  I mean... the whole point to this API is to wrap the builtin!  It should do the same thing, right?

File: include/toolchain.h [hunk 8]
```diff
@@ -15,6 +15,24 @@
...
+/**
+ * @def HAS_BUILTIN(x)
+ * @brief Check if the compiler supports the built-in function \a x.
+ *
+ * This macro is for use with conditional compilation to enable code using a
+ * builtin function that may not be available in every compiler.
+ */
+#ifdef __has_builtin
+#define HAS_BUILTIN(x) __has_builtin(x)
+#else
```
Comment (andyross) [inline include/toolchain.h]: Why bother here?  We're defining the HAS_BUILTIN_xxx symbols unconditionally anyway, right?  So what's the value of conditionally falling back to a toolchain macro when we could just do the same thing the same way for everyone and eliminate these lines?

File: kernel/include/syscall_handler.h [hunk 9]
```diff
@@ -353,9 +354,9 @@ extern int z_user_string_copy(char *dst, const char *src, size_t maxlen);
 #define Z_SYSCALL_MEMORY_ARRAY(ptr, nmemb, size, write) \
 	({ \
 		u32_t product; \
-		Z_SYSCALL_VERIFY_MSG(!__builtin_umul_overflow((u32_t)(nmemb), \
-							      (u32_t)(size), \
-							      &product), \
+		Z_SYSCALL_VERIFY_MSG(!u32_mul_overflow((u32_t)(nmemb), \
+						       (u32_t)(size), \
+						       &product), \
```
Comment (andyross) [inline kernel/include/syscall_handler.h]: Same deal.  Please segregate whitespace cleanup in separate commits.

File: kernel/poll.c [hunk 10]
```diff
@@ -273,11 +273,10 @@ Z_SYSCALL_HANDLER(k_poll, events, num_events, timeout)
...
 	}
-	if (Z_SYSCALL_VERIFY_MSG(
-		!__builtin_umul_overflow(num_events,
-					sizeof(struct k_poll_event),
-					&bounds),
-					"num_events too large")) {
+	if (Z_SYSCALL_VERIFY_MSG(!u32_mul_overflow(num_events,
+						   sizeof(struct k_poll_event),
+						   &bounds),
+				 "num_events too large")) {
```
Comment (andyross) [inline kernel/poll.c]: This change just looks like whitespace muckery unrelated to the patch.  Probably best to remove chaff from patches that actually do nontrivial stuff.

File: kernel/thread.c [hunk 11]
```diff
@@ -475,9 +476,8 @@ Z_SYSCALL_HANDLER(k_thread_create,
 	/* Verify that the stack size passed in is OK by computing the total
 	 * size and comparing it with the size value in the object metadata
 	 */
-	Z_OOPS(Z_SYSCALL_VERIFY_MSG(!__builtin_uadd_overflow(K_THREAD_STACK_RESERVED,
-							     stack_size,
-							     &total_size),
+	Z_OOPS(Z_SYSCALL_VERIFY_MSG(!u32_add_overflow(K_THREAD_STACK_RESERVED,
+						      stack_size, &total_size),
```
Comment (andyross) [inline kernel/thread.c]: More whitespace cleanup; should be in separate patch.

File: include/misc/math_extras_impl.h [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+
+	return a > 0 && c / a != b;
+}
+
+static inline int u32_count_leading_zeros(u32_t x)
+{
+	int b = 0;
+
+	while (b < 32 && (x >> 31) == 0) {
+		b++, x <<= 1;
```
Comment (andyross) [inline include/misc/math_extras_impl.h]: Same note about comma operator.

**Review by andyross — CHANGES_REQUESTED** [review]
I'm much more comfortable with the build integration, thanks.  A few style nitpicks that you can consider or ignore.

Just one question about the CTZ implementation is all I can find to whine about.

File: include/misc/math_extras_impl.h [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+	return a > 0 && c / a != b;
+}
+
+static inline bool size_mul_overflow(size_t a, size_t b, size_t *result)
+{
+	size_t c = a * b;
+
+	*result = c;
+
+	return a > 0 && c / a != b;
```
Comment (jukkar) [inline include/misc/math_extras_impl.h]: It would be really helpfull if you could add () here and other returns above. Now it needs mental effort to read the line in order to figure out what it is doing.

**Comment by olofj** [comment]
> The broader complaint is that we already have a framework for specifying toolchain configuration statically via ZEPHYR_TOOLCHAIN_VARIANT and the toolchain.h header tree, and we should be leveraging that if at all possible. I just don't see the need to try to detect this at runtime, do we even HAVE toolchains that support these functions piecewise? Seems like all we really care about is "gcc" and "other", and can just handle things the way we already do, no?

Not really, external toolchain can be of different versions, and different versions of GCC might or might not have these built-ins (depending on how recent it is). That was one of Jakob's original observations.

File: include/misc/math_extras_impl.h (hunk 7 above)
Comment (stoklund) [reply include/misc/math_extras_impl.h]: Here's how I look at it: The purpose of *this patch* is indeed to wrap compiler builtins and remove them from the general source code. However, looking beyond this PR, the purpose of the API in `math_extras.h` is to provide a set of commonly useful little math routines. For users of the API, it is not interesting whether they are implemented in portable C or with compiler builtins.

The x86 BSF/BSR instructions produce an undefined result for x=0, and GCC turned that into full-blown nasal demons C undefined behavior in their documentation of the ctz/clz builtins. Other instruction set architectures also have clz/ctz instructions, and they all agree that x=0 should return the size of the register. Intel even added LZCNT and TZCNT instructions that do the same.

So since there is a meaningful return value for x=0, I thought it would be better to return that instead of exposing users to nasal demons in a pretty surprising way.

Clang understands what's going on. Here's its armv7 code:

```asm
_u32_count_leading_zeros:
	clz	r0, r0
	bx	lr

_u32_count_trailing_zeros:
	rbit	r0, r0
	clz	r0, r0
	bx	lr
```

File: include/toolchain.h (hunk 8 above)
Comment (stoklund) [reply include/toolchain.h]: This case covers Clang-derived toolchains and [possibly future GCC versions](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=66970).

I'm trying to cover a more open-ended set of toolchains, a bit like Oticon's efforts.

File: include/misc/math_extras_impl.h (hunk 7 above)
Comment (andyross) [reply include/misc/math_extras_impl.h]: Seems like this needs an explanatory comment then about the inconsistent edge case, and maybe a `#ifdef __i386` or whatnot for clarity.   While it's nice that clang optimizes this correctly on ARM, I can all but guarantee you that many of our toolchains won't.

Comment (stoklund) [reply include/misc/math_extras_impl.h]: Unfortunately, `__builtin_ctz(0)` is undefined behavior on *all* architectures, thanks to the way it is defined and documented.

This means that if your code contains a line with `__builtin_ctz(x)` anywhere, the compiler is permitted to assume that x != 0 anywhere it can prove that this line has been or will be executed:

```c
  int first_missile = __builtin_ctz(x);

  if (x == 0) {
      disarm_missiles();
  }

  launch_missiles();
```

Here, the compiler is permitted to remove the conditional. It doesn't matter which instruction it ends up using for the builtin.

Many people find this infuriating, and understandably so.

**Comment by stoklund** [comment]
Updated patch:

- Clarify the precedence of arithmetic and relations with parentheses.
- Avoid the use of the comma operator and use a for loop instead of a while loop.
- Verified that all the patches are free of whitespace cleanup hunks.

File: include/misc/math_extras_impl.h (hunk 7 above)
Comment (andyross) [reply include/misc/math_extras_impl.h]: Then... maybe __builtin_ctz() is just problematic and a better solution would be arch-level inline assembly wrappers?  I mean... if the tool is broken why are we using it?

**Comment by mbolivar** [comment]
> FWIW, the cmake criticism was mostly aesthetics. I'm not really interested in litgating that per se, the point is that I'd prefer the lowest level of our headers be free of tool interactions. C code should be C code.

I find this comment frankly hilarious given the number of `CONFIG_` and `DT_` uses in our code.

File: include/misc/math_extras_impl.h (hunk 7 above)
Comment (olofj) [reply include/misc/math_extras_impl.h]: @andyross either way, isn't this patch a step in the right direction, in that it gives a single helper entrypoint such that if you want to replace it with an assembly stub later for a specific arch, you can? Meanwhile, using this instead of open-coding __builtin_foo() in the code is a clear incremental improvement.

Comment (andyross) [reply include/misc/math_extras_impl.h]: Fair enough.  I'll remove the -1.  Can we at least get a big comment on the usage here explaining that extra test and the fact that it's (hopefully) optimized away for the benefit of people like me would might be tempted to come in and "fix" it later?

**Review by andyross — APPROVED** [review]

**Comment by stoklund** [comment]
Thanks @andyross. I've pushed a rebased version which also adds a big comment about the clz/ctz zero checks.

**Review by tyrantlizard77 — APPROVED** [review]

**Review by galak — APPROVED** [review]
